// watson-developer-cloud/services/dialog/v1.js
/**
 * Copyright 2015 IBM Corp. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var extend         = require('extend');
var requestFactory = require('../../lib/requestwrapper');
var pick           = require('object.pick');
var omit           = require('object.omit');
var isStream       = require('isstream');

function Dialog(options) {
  // Default URL
  var serviceDefaults = {
    url: 'https://gateway.watsonplatform.net/dialog-beta/api'
  };

  // Replace default options with user provided
  this._options = extend(serviceDefaults, options);
}

/**
 * Retreives the values for profile variables for specific client ID
 * @param  Object   params   { client_id: '', dialog_id: '', name:''}
 */
Dialog.prototype.getProfile = function(params, callback) {
  params = params || {};

  var parameters = {
    options: {
      url: '/v1/dialogs/{dialog_id}/profile',
      method: 'GET',
      json: true,
      path: pick(params, ['dialog_id']),
      qs: omit(params, ['dialog_id'])
    },
    requiredParams: ['dialog_id', 'client_id'],
    defaultOptions: this._options
  };
  return requestFactory(parameters, callback);
};

/**
 * Sets the values for profile variables
 * @param  Object   params   { client_id: '', dialog_id: '', name:''}
 */
Dialog.prototype.updateProfile = function(params, callback) {
  params = params || {};

  if (!isStream(params.audio)) {
    callback(new Error('dialo is not a standard Node.js Stream'));
    return;
  }

  var parameters = {
    options: {
      url: '/v1/dialogs/{dialog_id}/profile',
      method: 'PUT',
      json: true,
      path: pick(params, ['dialog_id']),
      qs: omit(params, ['dialog_id'])
    },
    requiredParams: ['dialog_id', 'client_id'],
    defaultOptions: this._options
  };
  return requestFactory(parameters, callback);
};
module.exports = Dialog;
